vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs

```diff
@@ -26,89 +25,127 @@ open Microsoft.VisualStudio.Language.Intellisense
+        let navigateTo (range:range) = 
             asyncMaybe { 
-                let! id = documentId range
-                let! src = solution.GetDocument(id).GetTextAsync()
-                let! span = CommonRoslynHelpers.TryFSharpRangeToTextSpan(src, range)
-                if documentNavigationService.TryNavigateToSpan(workspace, id, span) then
-                   let! session = currentSession
-                   session.Dismiss()   
-            } |> Async.Ignore |> Async.StartImmediate 
-
-        let formatMap = typemap.ClassificationFormatMapService.GetClassificationFormatMap("tooltip")
-
-        let props = 
-            ClassificationTags.GetClassificationTypeName
+                let targetPath = range.FileName 
+                logInfof "origin document - %s \n"  initialDoc.FilePath
+                logInfof "target range to - %s \n %A"  range.FileName range
+                let! targetId = docIdOfRange range
+                let targetDoc = solution.GetDocument targetId 
+                let! targetSource = targetDoc.GetTextAsync() 
+                let! targetTextSpan = CommonRoslynHelpers.TryFSharpRangeToTextSpan (targetSource, range)
+
+                match isSignatureFile initialDoc.FilePath, isSignatureFile targetPath with 
+                | true, true 
+                | false, false ->
+                    return (gotoDefinitionService.TryNavigateToTextSpan (targetDoc, targetTextSpan))
+                // adjust the target from signature to implementation
+                | false, true ->
+                    return! gotoDefinitionService.NavigateToSymbolDefinitionAsync (targetDoc, targetSource, range)
+                // adjust the target from implmentation to signature
+                | true, false -> 
+                    return! gotoDefinitionService.NavigateToSymbolDeclarationAsync (targetDoc, targetSource, range)
+            } |> Async.map (Option.map (fun res -> 
+                if res then 
+                    SessionHandling.currentSession
+                    |> Option.iter (fun session -> session.Dismiss ())
+                )) |> Async.Ignore |> Async.StartImmediate 
+
+        let formatMap = typemap.ClassificationFormatMapService.GetClassificationFormatMap "tooltip"
+
+        let layoutTagToFormatting =
+            roslynTag
+            >> ClassificationTags.GetClassificationTypeName
             >> typemap.GetClassificationType
             >> formatMap.GetTextProperties
 
         let inlines = seq { 
-            for (tag, text, rangeOpt) in content do
-                let run =
-                    match rangeOpt with
-                    | Some(range) when canGoTo range ->
-                        let h = Documents.Hyperlink(Documents.Run(text), ToolTip = range.FileName)
-                        h.Click.Add <| fun _ -> goTo range          
+            for taggedText in content do
+                let run = Documents.Run(taggedText.Text, ToolTip = taggedText.Tag ) :> Documents.Inline
```

Comment:
Tooltip here was or debug, it shows LayoutTag for element at the moment. We can remove it for now.